src/rnnlm/rnnlm-lattice-rescoring.cc

```diff
@@ -0,0 +1,204 @@
+    const std::string &rnn_wordlist,
+    const std::string &word_symbol_table_rxfilename,
+    const RnnlmSimpleLoopedInfo &info) {
+  max_ngram_order_ = max_ngram_order;
+  ReadFstWordSymbolTableAndRnnWordlist(rnn_wordlist,
+                                       word_symbol_table_rxfilename);
+
+  std::vector<Label> bos_seq;
+  bos_seq.push_back(bos_index_);
+  state_to_wseq_.push_back(bos_seq);
+  RnnlmSimpleLooped *decodable_rnnlm = new RnnlmSimpleLooped(info);
+  decodable_rnnlm->TakeFeatures(bos_seq);
+  CuVector<BaseFloat> *hidden = decodable_rnnlm->GetOutput(0);
+  wseq_to_state_[bos_seq] = 0;
+  start_state_ = 0;
+
+  state_to_decodable_rnnlm_.push_back(decodable_rnnlm);
+  state_to_nnet3_output_.push_back(hidden);
+}
+
+fst::StdArc::Weight KaldiRnnlmDeterministicFst::Final(StateId s) {
+  // At this point, we should have created the state.
+  KALDI_ASSERT(static_cast<size_t>(s) < state_to_wseq_.size());
+
+  const CuVector<BaseFloat> &nnet3_out = *state_to_nnet3_output_[s];
+  RnnlmSimpleLooped* rnn = state_to_decodable_rnnlm_[s];
+  return rnn->LogProbOfWord(eos_index_, nnet3_out);
+}
+
+bool KaldiRnnlmDeterministicFst::GetArc(StateId s, Label ilabel,
+                                        fst::StdArc *oarc) {
+  // At this point, we should have created the state.
+  KALDI_ASSERT(static_cast<size_t>(s) < state_to_wseq_.size());
+
+  std::vector<Label> wseq = state_to_wseq_[s];
+//  RnnlmSimpleLooped* rnnlm = new RnnlmSimpleLooped(*state_to_decodable_rnnlm_[s]);
+  const RnnlmSimpleLooped* rnnlm = state_to_decodable_rnnlm_[s];
+  int32 rnn_word = fst_label_to_rnn_label_[ilabel];
+
+  const CuVector<BaseFloat> &nnet3_out = *state_to_nnet3_output_[s];
+  BaseFloat logprob = rnnlm->LogProbOfWord(rnn_word, nnet3_out);
+
+//  if (rnn_word == out_OOS_index_)
+//    logprob = logprob - Log(full_voc_size_ - rnn_label_to_word_.size() + 1.0);
+
+  wseq.push_back(rnn_word);
+  if (max_ngram_order_ > 0) {
+    while (wseq.size() >= max_ngram_order_) {
+      // History state has at most <max_ngram_order_> - 1 words in the state.
+      wseq.erase(wseq.begin(), wseq.begin() + 1);
+    }
+  }
+
+  std::pair<const std::vector<Label>, StateId> wseq_state_pair(
+      wseq, static_cast<Label>(state_to_wseq_.size()));
+
+  // Attemps to insert the current <lseq_state_pair>. If the pair already exists
+  // then it returns false.
+  typedef MapType::iterator IterType;
+  std::pair<IterType, bool> result = wseq_to_state_.insert(wseq_state_pair);
+
+  // If the pair was just inserted, then also add it to state_to_* structures
+  if (result.second == true) {
+    RnnlmSimpleLooped *rnnlm2 = new RnnlmSimpleLooped(*rnnlm);  // make a copy
+
+//  std::vector<int32> fst_label_to_rnn_in_label_;
+//  std::vector<std::string> rnn_in_label_to_word_;
+    rnnlm2->TakeFeatures(std::vector<Label>(1, rnn_word));
+    state_to_wseq_.push_back(wseq);
+    state_to_nnet3_output_.push_back(rnnlm2->GetOutput(0));
```

Comment:
I don't understand why this is zero here.
Either this is a bug, or your interfaces don't make sense.